File: Sorting_Techniques.java

public class Sorting_Techniques {

    public static void main(String[] args) {
        int[] nums = {4, 8, 3, 7, 1, 2, 5};
        int n = nums.length;
        // BubbleSort(nums, n);
        // SelectionSort(nums, n);
        // InsertionSort(nums);
        // MergeSort(nums);
        printArray(nums);
    }

    public static void BubbleSort(int[] nums, int n) {
        for (int i = 0; i < n - 1; i++) {
            boolean swapped = false;
            for (int j = 0; j < n - i - 1; j++) {
                if (nums[j] > nums[j + 1]) {
                    int temp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = temp;

                    swapped = true;
                }
            }

            if (!swapped) {
                return;
            }
        }
    }

    public static void SelectionSort(int[] nums, int n) {
        for (int i = 0; i < n - 1; i++) {
            int mpos = i;
            for (int j = i + 1; j < n; j++) {
                if (nums[j] < nums[mpos]) {
                    mpos = j;
                }
            }

            int temp = nums[mpos];
            nums[mpos] = nums[i];
            nums[i] = temp;
        }
    }

    private static void InsertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int j = i;
            while (j > 0 && arr[j] < arr[j - 1]) {
                int temp = arr[j];
                arr[j] = arr[j - 1];
                arr[j - 1] = temp;
                j--;
            }
        }
    }

    public static void recMergeSort(int[] theSeq, int left, int right) {
        if (left < right) {

            // Middle point
            int mid = left + (right - left) / 2;

            // Sort first and second halves
            recMergeSort(theSeq, left, mid);
            recMergeSort(theSeq, mid + 1, right);

            // Merge sorted halves
            merge(theSeq, left, mid, right);

        }

    }

    public static void MergeSort(int[] theSeq) {
        recMergeSort(theSeq, 0, theSeq.length - 1);
    }

    private static void merge(int[] theSeq, int left, int mid, int right) {
        int n1 = mid - left + 1;
        int n2 = right - mid;

        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++) {
            L[i] = theSeq[left + i];
        }

        for (int j = 0; j < n2; j++) {
            R[j] = theSeq[mid + 1 + j];
        }

        int i = 0, j = 0;
        int k = left;

        while (i < n1 && j < n2) {
            if (L[i] <= R[j]) {
                theSeq[k] = L[i];
                i++;
            } else {
                theSeq[k] = R[j];
                j++;
            }

            k++;
        }

        while (i < n1) {
            theSeq[k] = L[i];
            i++;
            k++;
        }

        while (j < n2) {
            theSeq[k] = R[j];
            j++;
            k++;
        }
    }

    public static int partitionSequence(int[] arr, int low, int high) {
        // Choose pivot 
        int pivot = arr[high];
        int i = low - 1;

        for (int j = low; j <= high - 1; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }

        swap(arr, i + 1, high);
        return i + 1;
    }

    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void quickSort(int[] theSeq, int first, int last) {
        // Partition Index: index of the pivot
        int pi = partitionSequence(theSeq, first, last);

        quickSort(theSeq, 0, pi - 1);
        quickSort(theSeq, pi + 1, last);
    }

    public static void printArray(int[] nums) {
        for (int i = 0; i < nums.length; i++) {
            System.out.print(nums[i] + " ");
        }
    }
}

File: Sorting_Techniques.py
## Bubble Sort
def Bubble_Sort(L):
    n = len(L)
    if n <= 1:
        return L
    for i in range(n - 1):
        for j in range(n - 1):
            if L[j] > L[j + 1]:
                L[j], L[j + 1] = L[j + 1], L[j]
    return L
## Time Complexity: O(n^2) 

## Selection Sort
def Selection_Sort(L):
    n = len(L)
    
    if n <= 1:
        return L
    
    for i in range(n):
        # Assume L[:i] is sorted
        mpos = i
        # mpos : position of minimum in L[i:]
        for j in range(i + 1, n):
            if L[j] < L[mpos]:
                mpos = j
            
        # L[mpos] : smallest value in L[i:]
        # Exchange L[mpos] and L[i]
        (L[i], L[mpos]) = (L[mpos], L[i])
        
        # Now L[:i+1] is sorted
    return (L)
## Time Complexity: O(n^2)

##Insertion Sort
def Insertion_Sort(L):
    n = len(L)
    if n <= 1:
        return L
    for i in range(n):
        # Assume L[:i] is sorted
        # Move L[i] to correct position in L
        j = i
        while(j > 0 and L[j] < L[j - 1]):
            (L[j], L[j - 1]) = (L[j - 1], L[j])
            j = j - 1
        # Now L[:i + 1] is sorted
    return (L)
## Time Complexity: O(n^2)

# New list created in recursive insertion sort
## Recursive Insertion Sort
def Insert(L, v):
    n = len(L)
    if n == 0:
        return [v]
    if v >= L[-1]:
        return (L + [v])
    else:
        return Insert(L[:-1], v) + L[-1:]

def ISort(L):
    n = len(L)
    if n <= 1:
        return (L)
    L = Insert(ISort(L[:-1]), L[-1])
    return (L)
## Time Complexity: O(n^2)

## Merge Sort
def Merge(A, B):
    (m, n) = (len(A), len(B))
    (C, i, j, k) = ([], 0, 0, 0)
    while k < m + n:
        if i == m:
            C.extend(B[j:])
            break
        elif j == n:
            C.extend(A[i:])
            break
        elif A[i] < B[j]:
            C.append(A[i])
            (i, k) = (i + 1, k + 1)
        else: 
            C.append(B[j])
            (j, k) = (j + 1, k + 1)
    return (C)

def Merge_Sort(L):
    n = len(L)
    
    if n <= 1:
        return (L)
    
    mid = n // 2
    L1 = Merge_Sort(L[:mid])
    L2 = Merge_Sort(L[mid:])
    sorted_L = Merge(L1, L2)
    
    return sorted_L
## Time Complexity: O(nlog(n))

## Quick Sort
def Quick_Sort(L, l, r):
    if (r - l <= 1):
        return L
    
    (pivot, lower, upper) = (L[l], l + 1, l + 1)
    for i in range(l + 1, r):
        if L[i] > pivot:
            upper += 1
        else: 
            (L[i], L[lower]) = (L[lower], L[i])
            (lower, upper) = (lower + 1, upper + 1)
    
    (L[l], L[lower - 1]) = (L[lower - 1], L[l])
    lower -= 1
    
    Quick_Sort(L, l, lower)
    Quick_Sort(L, lower + 1, upper)
    return L

L = [1, 2, 8, 9, 0, 3]

# print(Selection_Sort(L))
# print(Insertion_Sort(L))  
# print(ISort(L))  
# print(Merge_Sort([1, 2, 0, 8, -1, 9]))
# print(Bubble_Sort(L))
# print(Quick_Sort(L, 0, len(L)))